### PR TITLE
[GEO] 6x Deprecate ShapeBuilders and decouple geojson parse logic

### PR DESCRIPTION
**Pull request by nknize**
Like 7.0 this commit refactors the geo_shape parsing logic into its own package separate from the Shape builders and deprecates ShapeBuilders helper classes and methods. It also decouples the GeoShapeType into a standalone enumerator that is responsible for validating the parsed data and providing the appropriate builder. This future-proofs the code making it easier to maintain and not only add new shape types but support for new geo formats (e.g., WKT).

### CODE REVIEW

**Comment by nknize** [comment]
@cbuescher This is the 6.x version of the #27212 PR. There are only 2 deprecations. `ShapeBuilders` class, and `ShapeBuilder.parse`. I also updated the migration docs. The rest of it is the same as the 7.0 PR. Let me know if I missed anything else that needed to be deprecated. Thx!

File: core/src/test/java/org/elasticsearch/common/geo/BaseGeoParsingTestCase.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import com.vividsolutions.jts.geom.GeometryFactory;
+import org.elasticsearch.test.ESTestCase;
+
+import java.io.IOException;
+
+import static org.elasticsearch.common.geo.builders.ShapeBuilder.SPATIAL_CONTEXT;
+
+/**
+ * Created by nknize on 9/22/17.
+ */
```
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/common/geo/BaseGeoParsingTestCase.java]: Nit: only place I seen this so far, auto-created?

File: core/src/main/java/org/elasticsearch/common/geo/builders/ShapeBuilders.java [hunk 2]
```diff
@@ -26,7 +26,11 @@
 
 /**
  * A collection of static methods for creating ShapeBuilders.
+ *
+ * @deprecated this class is replaced by each builder's constructor
+ * e.g. instead of using {@code ShapeBuilders.newMultiPoint} use {@code new MultiPointBuilder()}
  */
+@Deprecated
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/geo/builders/ShapeBuilders.java]: `public static void register(List<Entry> namedWriteables)` is unused and can probably already be removed, should only be used internally.

File: core/src/main/java/org/elasticsearch/common/geo/builders/PointBuilder.java [hunk 3]
```diff
@@ -19,86 +19,78 @@
...
+    }
+
+    /**
+     * Create a new point
+     *
+     * @param longitude longitude of the point
+     * @param latitude latitude of the point
+     * @return a new {@link PointBuilder}
+     */
+    public static PointBuilder newPoint(double longitude, double latitude) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/geo/builders/PointBuilder.java]: Where does this come from? Seems unused.

File: core/src/main/java/org/elasticsearch/common/geo/builders/ShapeBuilder.java [hunk 4]
```diff
@@ -415,293 +386,57 @@ public static Orientation fromString(String orientation) {
...
-                } else {
-                    throw new ElasticsearchParseException("invalid LinearRing found (coordinates are not closed)");
-                }
-            }
-            return parseLineString(coordinates);
-        }
+    /** @deprecated this method will be removed in a future version; use ShapeParser.parse instead */
+    @Deprecated
+    public static ShapeBuilder parse(XContentParser parser) throws IOException {
+        DEPRECATION_LOGGER.deprecated("deprecated method [ShapeBuilder.parse] used; replaced by [ShapeParser.parse]");
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/geo/builders/ShapeBuilder.java]: I don't think we need deprecation logging here, I think marking the java API should be enough.

File: core/src/main/java/org/elasticsearch/common/geo/parsers/CoordinateNode.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    public final Coordinate coordinate;
+    public final List<CoordinateNode> children;
+
+    /**
+     * Creates a new leaf CoordinateNode
+     *
+     * @param coordinate
+     *            Coordinate for the Node
+     */
+    protected CoordinateNode(Coordinate coordinate) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/geo/parsers/CoordinateNode.java]: nit: can be package private (maybe I didn't see that in the master PR)

File: core/src/main/java/org/elasticsearch/common/geo/parsers/CoordinateNode.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        this.children = null;
+    }
+
+    /**
+     * Creates a new parent CoordinateNode
+     *
+     * @param children
+     *            Children of the Node
+     */
+    protected CoordinateNode(List<CoordinateNode> children) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/geo/parsers/CoordinateNode.java]: nit: can be package private (maybe I didn't see that in the master PR)

File: core/src/main/java/org/elasticsearch/common/geo/parsers/CoordinateNode.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        this.children = children;
+        this.coordinate = null;
+    }
+
+    public boolean isEmpty() {
+        return (coordinate == null && (children == null || children.isEmpty()));
+    }
+
+    public boolean isMultiPoint() {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/geo/parsers/CoordinateNode.java]: seems unused

**Review by cbuescher — COMMENTED** [review]
@nknize great, I left a few comments (mostly very minor nits, some might also be good to check on master), the deprecation looks good otherwise.

**Review by cbuescher — APPROVED** [review]
@nknize thanks for adding those changes, LGTM

**Pull request closed without merging**